File: src/extractors/sites/pixiv_extractor.cpp

#include <nlohmann/json.hpp>

#include "pixiv_extractor.hpp"

#include <curl/curl.h>

#include <iostream>
#include <regex>
#include <set>
#include "../../http/http_client.hpp"
#include "../../auth/cookie_manager.hpp"

using json =
    nlohmann::json;



bool PixivExtractor::matches(
    const std::string& url
)
{
    return
        url.find("pixiv.net")
        != std::string::npos;
}


std::string
PixivExtractor::extract_artwork_id(
    const std::string& url
)
{
    std::regex pattern(
        R"(/artworks/(\d+))"
    );

    std::smatch match;

    if (
        std::regex_search(
            url,
            match,
            pattern
        )
    )
    {
        return match[1];
    }

    return "";
}


std::string
PixivExtractor::http_get(
    const std::string& url
)
{
    HttpClient client;

    HttpRequest request;

    request.url = url;

    request.cookies =
        CookieManager::load(
            "pixiv"
        );

    request.referer =
        "https://www.pixiv.net/";

    request.headers =
    {
        "Referer: https://www.pixiv.net/"
    };

    HttpResponse response =
        client.get(request);

    if (response.status_code >= 400)
    {
        std::cerr
            << "Pixiv API error: HTTP "
            << response.status_code
            << " for "
            << url
            << "\n";

        return "";
    }

    return response.body;
}




std::vector<DownloadTask>
PixivExtractor::extract(
    const std::string& url
)
{
    std::string artwork_id =
        extract_artwork_id(url);


    if (artwork_id.empty())
    {
        std::cerr
            << "Could not extract artwork ID from URL\n";

        return {};
    }


    std::vector<DownloadTask> tasks;


    std::string info_url =
        "https://www.pixiv.net/ajax/illust/"
        + artwork_id;
    
    std::string info_json =
        http_get(info_url);

    if (info_json.empty())
    {
        std::cerr
            << "Pixiv illust info: empty response\n";

        return {};
    }

    json info_parsed;

    try
    {
        info_parsed =
            json::parse(
                info_json
            );
    }

    catch (const std::exception& e)
    {
        std::cerr
            << "Pixiv illust info parse error: "
            << e.what()
            << "\n";

        return {};
    }


    bool is_ugoira = false;
    
    if (
        info_parsed.contains("body")
    )
    {
        const auto& body =
            info_parsed["body"];
    
        if (
            body.contains("illustType")
        )
        {
            int illust_type =
                body["illustType"];

            if (illust_type == 2)
            {
                is_ugoira = true;
            }
        }
    }


    if (is_ugoira)
    {
        std::string ugoira_url =
            "https://www.pixiv.net/ajax/illust/"
            + artwork_id +
            "/ugoira_meta";

        std::string ugoira_json =
            http_get(ugoira_url);

        if (ugoira_json.empty())
        {
            std::cerr
                << "Failed to load ugoira metadata\n";

            return {};
        }

    json ugoira_parsed;

    try
    {
        ugoira_parsed =
            json::parse(
                ugoira_json
            );
    }

    catch (const std::exception& e)
    {
        std::cerr
            << "Ugoira metadata parse error: "
            << e.what()
            << "\n";

        return {};
    }

    if (
        !ugoira_parsed.contains(
            "body"
        )
    )
    {
        std::cerr
            << "Ugoira metadata: missing 'body'\n";

        return {};
    }

    const auto& body =
        ugoira_parsed["body"];

    if (
        !body.contains(
            "originalSrc"
        )
    )
    {
        std::cerr
            << "Ugoira metadata: missing 'originalSrc'\n";

        return {};
    }

    std::string zip_url =
        body["originalSrc"];

    DownloadTask task;

    task.url =
        zip_url;

    task.headers =
    {
        "Referer: https://www.pixiv.net/"
    };

    task.cookies =
        "PHPSESSID=113246217_AYi4n9iGcpKKTlMcy9nCp9v25pQS7HNr";

    task.filename =
        artwork_id
        +
        "_ugoira.zip";

    if (
        body.contains("frames")
    )
    {
        for (const auto& frame : body["frames"])
        {
            if (frame.contains("delay"))
            {
                task.frame_delays.push_back(
                    frame["delay"]
                );
            }
        }
    }

    tasks.push_back(task);

    return tasks;


    }




    std::string api_url =
        "https://www.pixiv.net/ajax/illust/"
        + artwork_id +
        "/pages";

    std::string json_data =
        http_get(api_url);

    if (json_data.empty())
    {
        std::cerr
            << "Pixiv pages: empty response\n";

        return {};
    }





    json parsed;
    
    try
    {
        parsed =
            json::parse(
                json_data
            );
    }
    catch (const std::exception& e)
    {
        std::cerr
            << "Pixiv pages parse error: "
            << e.what()
            << "\n";

        return {};
    }
    
    
    if (!parsed.contains("body"))
    {
        return {};
    }
    
    for (const auto& page : parsed["body"])
    {
        if (
            !page.contains("urls")
            ||
            !page["urls"].contains("original")
        )
        {
            continue;
        }
    
        std::string original_url =
            page["urls"]["original"];
    
        DownloadTask task;
    
        task.url = original_url;
    
        task.headers =
        {
            "Referer: https://www.pixiv.net/"
        };
    
        task.cookies =
            "PHPSESSID=113246217_AYi4n9iGcpKKTlMcy9nCp9v25pQS7HNr";
    
        size_t last_slash =
            original_url.find_last_of('/');
    
        if (last_slash != std::string::npos)
        {
            task.filename =
                original_url.substr(
                    last_slash + 1
                );
        }
    
        tasks.push_back(task);
    }

    return tasks;
}

File: android/app/src/main/java/com/mediadownloader/extractors/sites/PixivExtractor.kt
package com.mediadownloader.extractors.sites

import com.mediadownloader.auth.CookieManager
import com.mediadownloader.extractors.BaseExtractor
import com.mediadownloader.extractors.DownloadTask
import com.mediadownloader.http.HttpClient
import com.mediadownloader.http.HttpRequest
import kotlinx.serialization.json.*

class PixivExtractor : BaseExtractor {
    private val json = Json { ignoreUnknownKeys = true }

    override fun matches(url: String): Boolean =
        url.contains("pixiv.net")

    override suspend fun extract(url: String): List<DownloadTask> {
        val artworkId = extractArtworkId(url)
        if (artworkId.isEmpty()) return emptyList()

        val cookies = CookieManager.load("pixiv")

        val infoBody = fetchJson(
            "https://www.pixiv.net/ajax/illust/$artworkId",
            cookies
        ) ?: return emptyList()

        val body = infoBody["body"]?.jsonObject ?: return emptyList()
        val illustType = body["illustType"]?.jsonPrimitive?.int ?: 0

        if (illustType == 2) {
            return extractUgoira(artworkId, cookies)
        }

        return extractPages(artworkId, cookies)
    }

    private suspend fun extractUgoira(
        artworkId: String,
        cookies: String
    ): List<DownloadTask> {
        val metaBody = fetchJson(
            "https://www.pixiv.net/ajax/illust/$artworkId/ugoira_meta",
            cookies
        ) ?: return emptyList()

        val body = metaBody["body"]?.jsonObject ?: return emptyList()
        val zipUrl = body["originalSrc"]?.jsonPrimitive?.content ?: return emptyList()

        val frameDelays = mutableListOf<Int>()
        for (frame in body["frames"]?.jsonArray ?: emptyList()) {
            frame.jsonObject["delay"]?.jsonPrimitive?.int?.let { frameDelays.add(it) }
        }

        return listOf(
            DownloadTask(
                url = zipUrl,
                filename = "${artworkId}_ugoira.zip",
                headers = mapOf("Referer" to "https://www.pixiv.net/"),
                cookies = cookies,
                frameDelays = frameDelays
            )
        )
    }

    private suspend fun extractPages(
        artworkId: String,
        cookies: String
    ): List<DownloadTask> {
        val pagesBody = fetchJson(
            "https://www.pixiv.net/ajax/illust/$artworkId/pages",
            cookies
        ) ?: return emptyList()

        val pages = pagesBody["body"]?.jsonArray ?: return emptyList()

        return pages.mapNotNull { page ->
            val originalUrl = page.jsonObject["urls"]
                ?.jsonObject
                ?.get("original")
                ?.jsonPrimitive
                ?.content ?: return@mapNotNull null

            DownloadTask(
                url = originalUrl,
                filename = originalUrl.substringAfterLast('/'),
                headers = mapOf("Referer" to "https://www.pixiv.net/"),
                cookies = cookies
            )
        }
    }

    private suspend fun fetchJson(
        url: String,
        cookies: String
    ): JsonObject? {
        val resp = HttpClient.get(
            HttpRequest(
                url = url,
                userAgent = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/136.0.0.0 Safari/537.36",
                referer = "https://www.pixiv.net/",
                headers = mapOf("Referer" to "https://www.pixiv.net/"),
                cookies = cookies
            )
        )
        if (resp.statusCode >= 400) return null

        return try {
            json.parseToJsonElement(resp.body).jsonObject
        } catch (_: Exception) {
            null
        }
    }

    private fun extractArtworkId(url: String): String {
        val pattern = Regex("""/artworks/(\d+)""")
        return pattern.find(url)?.groupValues?.get(1) ?: ""
    }
}
